Add command history and line editing to ocamldebug

joelreymont opened the pull request:
Implement readline-like functionality with arrow key navigation, tab completion, inline hints, and persistent history. Integrate with the debugger event loop for non-blocking operation.

Features:
- History navigation (Up/Down arrows)
- Cursor movement (Left/Right, Home/End, Ctrl+A/E)
- Line editing (Backspace, Delete, Ctrl+K)
- Tab completion (commands, variables, info topics)
- Inline syntax hints
- Persistent history (~/.ocamldebug_history)
- Automatic TTY detection with fallback
- Ctrl+C cancels current line without exiting debugger

Compatibility:
- Unix/Linux/macOS via termios
- Windows via automatic fallback to legacy input
- Pipe/script mode unchanged
- Emacs and machine-readable modes supported
- Respects -no-prompt flag

gasche commented:
Hi @joelreymont,

Your previous PR today #14350 turned out to be a waste of time to review, as it contained some relatively simple programming mistakes which should probably have caught by self-review, and in fact it implements a feature that was already available. These things happen, but to me it suggests that some sort of quality-control on your side was missing, and before I consider spending my review time by looking at this one I would like to be sure that you changed your process to fix this issue.

Have you read carefully the patch that you are submitting? Could you consider splitting the change into separate commits that each contain only the specific code implementing a specific feature?

My overall impression from a distance is that these features are useful, but also something that could gather a lot of changes due to for example keybinding tastes and other UI feature ideas, and so could be a strain maintenance-wise. I wonder if we should consider having it out of tree as a separate program, providing that the base `ocamldebug` UI provides the flexibility to do so. (I suppose it does as the Emacs mode provides these features, but I'm not sure.)

joelreymont commented:
> Have you read carefully the patch that you are submitting? 

I actually tested it to make sure it works. Further, I spent half a day fixing an issue. 

The code looks reasonably clean to me. Tests are written in the compiler style, as opposed to separate shell scripts. 

> Could you consider splitting the change into separate commits that each contain only the specific code implementing a specific feature?

I could try. There are only 3 features here and they are 
- Arrow navigation up, down, left and right, as well as Ctrl-A and Ctrl-E for beginning and end of line respectively.
 - Tab completion.
 - Managing history.
 
Do you want me to split the commits accordingly?

> My overall impression from a distance is that these features are useful, but also something that could gather a lot of changes due to for example keybinding tastes and other UI feature ideas, and so could be a strain maintenance-wise. 

I believe these keybindings are quite universal. It's just arrows after all, plus a couple of other keys. They are supported and provided by `libreadline` used in `lldb`, `gdb`, etc. 

My implementation is self-standing and depends on no other external libraries. It also comes with a test suite. 

> I wonder if we should consider having it out of tree as a separate program, providing that the base `ocamldebug` UI provides the flexibility to do so. (I suppose it does as the Emacs mode provides these features, but I'm not sure.)

That program already exists and is called `rlwrap`. It does not provide tab completion and comes with some inconveniences which is why `lldb`, `gdb`, etc. use `libreadline` instead.

joelreymont commented:
I will try to create a commit by feature and check again for optimization opportunities and such.

joelreymont committed:
ocamldebug: use static buffer for character input

Replace per-keystroke Bytes.create with a single static buffer,
eliminating an allocation on every character typed during line editing.

joelreymont committed:
ocamldebug: cache hints to avoid recomputation

Add Hashtbl cache for hint computations, which are called on every
keystroke. The hints callback parses command syntax via the lexer,
making it expensive. Caching improves performance by ~10x for
subsequent keystrokes on the same partial command.

joelreymont committed:
ocamldebug: add termios and pty plumbing for line editing

Introduce basic infrastructure for character-by-character terminal
input handling:

- TTY detection and termios setup/restore for raw mode
- Character-mode callback support in Input_handling event loop
- command_history module with line_result type for processing
- Integration with main debugger initialization
- Automatic fallback to line mode when not on TTY or in emacs/machine modes

This commit provides the plumbing without implementing actual line
editing - that will come in subsequent feature commits.

joelreymont committed:
ocamldebug: add keyboard navigation and line editing

Implement interactive line editing with history navigation:

Line Editing:
- Character insertion/deletion at cursor position
- Cursor movement (Left/Right, Home/End, Ctrl+A/E)
- Backspace, Delete, Ctrl+K (kill to end)
- Escape sequence parsing for arrow keys and special keys
- ANSI terminal control for display updates

History Navigation:
- Up/Down arrows to navigate command history
- History storage with duplicate removal
- Configurable history size via "set history" command
- Automatic history tracking for executed commands

Signal Handling:
- Ctrl+C cancels current line without exiting
- Ctrl+Z and Ctrl+\ trigger terminal signals properly
- Ctrl+D on empty line triggers EOF

Integration:
- Seamless integration with Input_handling event loop
- Character-by-character callback processing
- Automatic fallback when TTY not available
- State machine for robust escape sequence handling

This provides a readline-like experience while maintaining
compatibility with non-interactive modes. Tab completion
support will be added in a subsequent commit.

joelreymont committed:
ocamldebug: add tab completion and inline hints

Implement intelligent tab completion for debugger commands:

Tab Completion:
- Command name completion (step, next, run, etc.)
- Variable name completion for "set" and "show" commands
- Info topic completion for "info" command
- Help command completion
- Context-aware multi-word completion
- Automatic single-match insertion
- Multiple-match display with redisplay

Inline Hints:
- Gray-text hints showing command syntax
- First line of help text displayed after command
- Non-intrusive, positioned after cursor
- Updates dynamically as you type

Implementation:
- Completion callback system for extensibility
- Hints callback for syntax assistance
- Word boundary detection for accurate replacement
- Temporary terminal mode switch for completion display
- Recursive editor restart after showing completions

Integration:
- Callback registration in command_line init
- Automatic command/variable/info matching
- Fallback to empty list on parse errors

Note: The completion display currently has a bug where it
restarts line editing and clears the buffer. This will be
fixed in optimization commits.

joelreymont committed:
ocamldebug: add persistent command history

Implement history persistence across debugger sessions:

History File:
- Default location: ~/.ocamldebug_history
- Fallback to local .ocamldebug_history if HOME not set
- Automatic creation on first save

Loading:
- Load history on debugger startup in init()
- Read from oldest to newest (reversed during load)
- Respect max_history_size configuration
- Silent failure if file doesn't exist or can't be read

Saving:
- Save history on debugger exit (via at_exit in main.ml)
- Write from oldest to newest for consistent format
- Only save on TTY to avoid pollution from scripts
- Atomic write with exception handling

Integration:
- Uses Debugger_config.history_size for default size
- Respects user's set history command
- Trims to max size after loading
- Duplicate removal maintained in add_to_history

File Format:
- One command per line
- Plain text, oldest first
- No special escaping (commands shouldn't contain newlines)

This provides a familiar readline-like experience where
command history persists between sessions, similar to
bash_history or .python_history.

joelreymont committed:
ocamldebug: add testing infrastructure and documentation

Add comprehensive test suite and documentation for line editing:

Test Suite (testsuite/tests/tool-debugger/history/):
- basic.ml: Tests basic history commands and navigation
- machine_readable.ml: Tests that machine mode disables line editing
- tty_history.ml: Tests interactive TTY behavior with pseudo-terminal
- tty_driver.ml: PTY driver for simulating interactive sessions
- tty_stubs.c: C stubs for PTY operations
- Reference files for expected output validation

Test Infrastructure Updates:
- ocamltest/debugger_actions.ml: Support for debugger_flags variable
- ocamltest/debugger_variables.ml/mli: New debugger_flags variable
- Allows tests to pass additional flags like -no-prompt

Man Page Documentation (man/ocamldebug.1):
- New "INTERACTIVE LINE EDITING" section
- Documents cursor movement, completion, and hints
- Explains signal handling (Ctrl+Z, Ctrl+\)
- Notes UTF-8 limitations and fallback behavior
- Describes automatic fallback for non-TTY modes

Test Coverage:
- History command execution and persistence
- Machine-readable mode compatibility
- TTY detection and line editing activation
- Prompt display and command execution
- All tests pass successfully

This provides a solid foundation for validating the line editing
functionality and documenting it for users.

joelreymont committed:
ocamldebug: keep editor active during tab completion

Fix critical bug where multi-match tab completion would wipe the
user's input buffer.

Problem:
When Tab produced multiple matches, show_completion_choices would:
1. Exit the character-mode callback loop
2. Restore terminal to canonical mode
3. Print completions
4. Recursively call read_from_editor()
5. Call start_line() which cleared current_line_buffer

Result: Typing "info<Tab>" would submit "modules" instead of
"info modules", making multi-match completion unusable.

Solution:
- Remove pending_completions variable and recursive call
- Call show_completion_choices immediately in callback
- Return true to continue the loop (don't exit)
- Keep terminal in raw mode during completion display
- Redisplay the current line after showing matches
- Remove 'rec' keyword from read_from_editor

This keeps the gap buffer intact and preserves the user's
partially-typed command when showing completion choices.

joelreymont committed:
ocamldebug: fix PTY tests for Linux

Fix build failures on Linux for the PTY-backed regression tests.

Problem 1: Wrong header file
- Linux exposes openpty via <pty.h>
- macOS uses <util.h>
- Test unconditionally included <util.h>

Problem 2: Missing link library
- glibc requires -lutil at link time for openpty
- Test description never supplied cclib flags

Solution:
- Conditional include: <pty.h> on Linux, <util.h> elsewhere
- Add cclib = "-lutil" to tty_driver.exe build step

This allows the history test suite to build and run on standard
Linux toolchains with glibc.

joelreymont committed:
ocamldebug: honor -no-prompt flag to disable line editing

Fix inconsistency where -no-prompt would still enable raw mode
editing.

Problem:
The man page promises "line editing automatically falls back
when prompts are disabled (-no-prompt)" but only -emacs and
-machine-readable actually disabled line editing in main.ml.

Running ocamldebug -no-prompt would still enter raw mode with
no visible prompt, contradicting the documented behavior.

Solution:
Add !Parameters.prompt check to set_enabled condition in main.ml,
so line editing is disabled when prompts are turned off.

This ensures the documented fallback behavior matches the actual
implementation.

joelreymont committed:
ocamldebug: fix double prompt printing

The debugger was printing two prompts for every command: once from
resume_user_input() registering stdin, and again when the line editor
started and displayed its prompt.

Fix by removing prompt printing from resume_user_input() and moving it
to command_line.ml only in the fallback case when line editing fails
to start (not a TTY or disabled). When line editing is active,
start_line() handles printing the prompt itself.

joelreymont committed:
ocamldebug: allow set history 0 for privacy

Users who want to disable history persistence for privacy reasons
can now set history to 0. Previously the minimum was enforced as 1,
which meant the last command would always be saved.

Changed validation from "< 1" to "< 0" to allow zero as a valid value.

joelreymont committed:
ocamldebug: fix PTY test for cross-platform compatibility

The tty_history test needs -lutil on Linux but not on macOS.
Instead of using invalid TSL syntax (cclib = "-lutil"), use a
shell script that conditionally adds the flag based on platform.

- Created build_tty_driver.sh that checks uname and adds -lutil on Linux
- Updated tty_history.ml to use the build script
- Test now passes on macOS and should work on Linux

This fixes the test infrastructure to properly handle platform-specific
linker requirements for openpty().

hhugo commented on debugger/command_history.ml:
```diff
@@ -0,0 +1,549 @@
+        | [] -> List.rev acc
+        | x :: xs -> take (n - 1) xs (x :: acc)
+    in
+    history := take limit !history []
```
```suggestion
    history := List.take limit !history
```

joelreymont committed:
ocamldebug: fix tty_history test script variable substitution

The build_tty_driver.sh script relied on ${ocamlc} being substituted,
but TSL variable substitution only occurs in script strings, not inside
shell scripts they invoke.

Fix by passing ${ocamlc} as an argument to the script, allowing TSL
to substitute it at the call site.

joelreymont committed:
ocamldebug: use List.take instead of custom implementation

Replace custom recursive take function with standard library List.take.

hhugo commented on debugger/command_history.ml:
```diff
@@ -0,0 +1,549 @@
+          with End_of_file ->
+            acc  (* Don't reverse - file has oldest first, :: builds newest first *)
+        in
+        history := load_lines [];
```
```suggestion
        history := In_channel.fold_lines (fun acc line -> line :: acc) [] ic;
```

joelreymont committed:
ocamldebug: use In_channel.fold_lines for history loading

joelreymont committed:
ocamldebug: fix tab completion trailing space and mid-word bugs

Two completion bugs fixed:

1. Trailing space: Typing 'set <TAB>' was treated as single-word
   command because string_trim removed the trailing space before
   tokenization. Now preserve trailing whitespace by appending empty
   token when cursor follows whitespace.

2. Mid-word completion: Completing 'breakp|oint' to 'breakpoint'
   produced 'breakpointeint' because suffix included remainder of
   original word. Now skip to word end before taking suffix.

hhugo commented on debugger/input_handling.ml:
```diff
@@ -89,6 +89,31 @@ let current_prompt = ref ""
+
+(* Controller for character-mode input *)
+let char_mode_controller iochan =
+  let n = input iochan.io_in char_input_buffer 0 1 in
```
In_channel.input_char ?

joelreymont committed:
ocamldebug: use In_channel functions for character and file I/O

Use In_channel.input_char for reading characters and In_channel.with_open_text
for history file loading to leverage standard library abstractions.

dbuenzli commented:
> I believe these keybindings are quite universal. It's just arrows after all, plus a couple of other keys. They are supported and provided by `libreadline` used in `lldb`, `gdb`, etc.

Note that in practice stuff using `libreadline` can be overridden by users using `~/.inputrc` though in `down` no-one complained loudly that I'm [still not doing](https://github.com/dbuenzli/down/issues/12) this.

> * Persistent history (~/.ocamldebug_history)

Nowadays it is regarded as poor style to write at the toplevel. I think this should be written to either `$XDG_CONFIG_HOME/ocaml/ocamldebug_history` or `$XDG_STATE_HOME/ocaml/ocamldebug_history`.

joelreymont commented:
> > * Persistent history (~/.ocamldebug_history)
> 
> Nowadays it is regarded as poor style to write at the toplevel. I think this should be written to either `$XDG_CONFIG_HOME/ocaml/ocamldebug_history` or `$XDG_STATE_HOME/ocaml/ocamldebug_history`.

Let me fix this! I suggest only writing to the top level if `$XDG_CONFIG_HOME` or `$XDG_STATE_HOME` are unset.

dbuenzli commented:
> Let me fix this! I suggest only writing to the top level if `$XDG_CONFIG_HOME` or `$XDG_STATE_HOME` are unset.

That's not the way the [XDG base specification](https://specifications.freedesktop.org/basedir/latest/) works, there are specific defaults and a choice needs to be done between these two directories. 

I don't know which one is more appropriate, in `down` I used `XDG_CONFIG_HOME/ocaml/history.ml` because `XDG_STATE_HOME` didn't exist at the time. However nowadays I think I would still be hesitant to use `XDG_STATE_HOME` because `down` also has session files which are saved history files and I wouldn't see these in `XDG_STATE_HOME`.

Also usually you have to do something sensitive on windows. IIRC something like this has been done for the lookup of `.ocamlinit` so perhaps look there.

dra27 commented:
The features are unquestionably useful, but I’m not sure that code-first is the best approach here. The basic debugger sources clearly need to live in the compiler codebase because they’re quite tied to details of the compiler (although it doesn't hugely change between versions - I think 5.0 was the first time in a while that considerable work had to be done to it), but carving our own _internal_ tui code seems excessive, on first glance.

I would consider exploring what can be done if the debugger were built completely after the compiler (i.e. as a separate package, just using the same source tree). That separation could provide the ability to plug _external_ (OCaml) libraries into this. utop is an example of a project doing this - hugely reliant and tied to the toplevel library, but utilising a suite of libraries to implement its own ux on top of that. In terms of how such a concept would live in the compiler codebase, ocamltest for example is capable of being built with or without the Unix library.

More simply, are there hooks and mechanisms which could be added to ocamldebug to allow an approach as in @dbuenzli’s down (similarly, we’ve added hooks in the past to the toplevel to allow things like the ocaml-jit project to work - external libraries plugged into custom toplevels, but without having to pull huge libraries into ocaml/ocaml)

joelreymont committed:
ocamldebug: use XDG_CONFIG_HOME for history file location

Follow XDG Base Directory specification using XDG_CONFIG_HOME,
consistent with OCaml's .ocamlinit lookup pattern.

Checks existing files in order before selecting location:
1. $XDG_CONFIG_HOME/ocaml/ocamldebug_history (if set and absolute)
2. $HOME/.config/ocaml/ocamldebug_history (XDG default)
3. $HOME/.ocamldebug_history (legacy fallback)
4. .ocamldebug_history (current directory)

Uses first existing file for loading history. For new history, uses
preferred XDG location. Creates parent directories automatically.

joelreymont committed:
ocamldebug: fix help completion and history line preservation

Two bugs fixed:

1. Help completion: Tab after 'help ' now completes instruction names.
   Previously only worked for single-word 'help' command.

2. History navigation: Preserve partially-typed line when browsing history.
   Arrow up/down now saves and restores working line instead of losing it.

joelreymont commented:
My concern is that the current `ocamldebug` experience is crap. 

I have already implemented keyboard navigation, command history and tab completion which solves the issues I'm experiencing. The change is completely self-contained and properly checked by automated tests, including `pty` functionality.

What about merging this PR as a stop-gap measure and _then_ kicking off a design discussion of how to properly improve the `ocamldebug` experience?

dbuenzli commented:
Another angle to the discussion is how central `ocamldebug` is deemed to the experience of the language. It seems to me that efforts are rather pushed towards native code source level debugging. 

It's also a bit unclear to me how much what is being proposed here overlaps with the [ide integration](https://ocaml.org/manual/5.4/debugger.html#s%3Ainf-debugger).

joelreymont commented:
This should be fixed to match `.ocamlinit` now.

joelreymont commented:
> Another angle to the discussion is how central `ocamldebug` is deemed to the experience of the language. It seems to me that efforts are rather pushed towards native code source level debugging.

There's no DWARF debugging support on macOS so I'm forced to use `ocamldebug`. I have a HUGE DWARF PR in the works but I'm still testing it.
 
> It's also a bit unclear to me how much what is being proposed here overlaps with the [ide integration](https://ocaml.org/manual/5.4/debugger.html#s%3Ainf-debugger).

Emacs support is preserved by this PR. Also, I don't use Emacs and use Neovim instead.

joelreymont committed:
ocamldebug: fix multi-word completion with trailing whitespace

Handle both rest = [] and rest = [""] cases when completing multi-word
commands (set, show, info, help) after whitespace. This ensures that
commands like "help <TAB>", "set <TAB>", etc. show completions correctly.

lthls commented:
> What about merging this PR as a stop-gap measure and _then_ kicking off a design discussion of how to properly improve the `ocamldebug` experience?

Why do you need this PR to get merged ? Even if it is, it won't be part of a release until several months, so if you want to use the features you're working on, it would be better to just create a new debugger project, copying the current sources with your modifications, that you could release immediately and where you wouldn't depend on other people to review your code (reviewer time is very precious around here).

You mention a stop-gap measure, but we already have such measures: combining ocamldebug with either `rlwrap`, `ledit`, or the Emacs mode will provide the features proposed here. So on our side there is no urgent bug to fix, and the urgency itself is suspicious.

Finally, when @gasche asked you if you had read the code, you answered that you had tested it. That's not the same. Asking people to review code that you haven't even read yourself feels disrespectful to me, and I suspect this will contribute to this PR getting stalled.

> There's no DWARF debugging support on macOS so I'm forced to use `ocamldebug`. I have a HUGE DWARF PR in the works but I'm still testing it.

I suggest looking in this project's pull requests tab at the ones labelled `debugging-prerequisite`. You will see that someone already proposed a huge DWARF PR, actually split it in a way suitable for reviewing, and it still got turned down (even though some parts were merged). If you don't have an argument showing why your approach is better you will likely end up wasting your time working on that.

dra27 commented:
> What about merging this PR as a stop-gap measure and _then_ kicking off a design discussion of how to properly improve the `ocamldebug` experience?

I'm afraid this is absurd, and completely ignores both review time and, more importantly, long-term maintenance. @lthls has already beaten me to mentioning the fairly obvious other stop-gap solutions - I was about to ask if `rlwrap ocamldebug` is particularly limiting here?

joelreymont commented:
> Finally, when @gasche asked you if you had read the code, you answered that you had tested it. That's not the same. Asking people to review code that you haven't even read yourself feels disrespectful to me, and I suspect this will contribute to this PR getting stalled.

I read the code and I re-read the code. I even split the code into many small commits for easier reading.
 
> I suggest looking in this project's pull requests tab at the ones labelled `debugging-prerequisite`. You will see that someone already proposed a huge DWARF PR, actually split it in a way suitable for reviewing, and it still got turned down (even though some parts were merged). If you don't have an argument showing why your approach is better you will likely end up wasting your time working on that.

Will do, thanks for the pointer! I [asked about it in the forum](https://discuss.ocaml.org/t/dwarf-support-on-macos/17474) but got no replies.

joelreymont commented:
> I suggest looking in this project's pull requests tab at the ones labelled `debugging-prerequisite`. You will see that someone already proposed a huge DWARF PR, actually split it in a way suitable for reviewing, and it still got turned down (even though some parts were merged). 

Are you talking about PR #2290 and PR #2316?

joelreymont commented:
> Are you talking about PR #2290 and PR #2316?

I only see comments about sub-optimal code. Did you have something else in mind @lthls ?

lthls commented:
The main PR is #574. I think the github comments will likely not reveal the true reason behind the PR being closed (the review process was fairly long, parts of it involved people meeting in person to discuss things, and the final decision to stop working on that occured in such a meeting, I believe).
The reason at the time was simple: it required too much time for review, which we (the core maintainers) wanted to spend on other things. The reason is still likely to be valid.

joelreymont commented:
Folks,

I respect your time, concerns and reasoning. 

I also have a better understanding of why Jane St started their fork of OCaml and is maintaining it.

Please feel free to close this PR if it has no chance of making it into OCaml.

joelreymont committed:
ocamldebug: refactor to eliminate code duplication

Extract repeated patterns into reusable helper functions:

1. command_line.ml:
   - Add trailing_token helper to extract last token from word list,
     eliminating triplicate logic in set/show/info/help completion
   - Add is_whitespace helper used in tokenization and trailing space check
   - Simplify multi-word completion by hoisting ident extraction

2. command_history.ml:
   - Extract mkdir_p to module level for reuse in future features
   - Add is_whitespace helper for word boundary detection
   - Add split_buffer_at helper to reduce duplication in insert/delete ops
   - Cache buffer references and lengths in completion handler

These changes improve maintainability and eliminate redundant List.rev
calls and repeated string operations.

dra27 commented:
> I also have a better understanding of why Jane St started their fork of OCaml and is maintaining it.

This is unnecessary (and the implication you're making is also wildly inaccurate, given the first paragraph [of the OxCaml landing page](https://oxcaml.org)).

joelreymont commented:
> This is unnecessary (and the implication you're making is also wildly inaccurate, given the first paragraph [of the OxCaml landing page](https://oxcaml.org)).

The only thing I'm implying (nay saying directly) is that it's easier to fork and maintain your own fork given the limited review bandwidth of OCaml maintainers and the desire to keep OCaml small enough to be maintainable. 

OCaml is has been forkable for a while now so there's no reason not to fork ocamldebug, ocaml compiler, etc. and then maintain it yourself. It does split the ecosystem but also keeps a certain balance between available resources and the needs of the users.

tmcgilchrist commented:
> Will do, thanks for the pointer! I [asked about it in the forum](https://discuss.ocaml.org/t/dwarf-support-on-macos/17474) but got no replies.

Gotta give me more time to respond than 1 day :-) 

There is significant existing work in flight on [OxCaml](https://github.com/oxcaml/oxcaml), [LLDB](https://github.com/ocaml-flambda/llvm-project/pulls), and [GDB](https://github.com/tmcgilchrist/binutils-gdb/pull/1) to support DWARF 5 and source debugging. Speaking for myself as an active participant (not for the Jane Street folks also working on this), I want to upstream this DWARF work next year, however the work is detailed and the source diff will be large. So upstreaming will take some time. 

If you want to improve bytecode debugging experience, in my opinion I would look at DAP integration in ocamlearlybird and focus on using DAP as the way to provide a UI for bytecode debugging. Then any modifications that need to happen on ocamldebug like making it work better with DAP would be well motivated. If you want a TUI that speaks DAP then build that as a UI on top of an ocamldebug core.

matkoniecz commented:
I wonder if this is an AI-generated pull request, without human review. This is based on their other pull requests to many different projects.

See say https://github.com/ocaml/ocaml/pull/14369#issuecomment-3557328285

dra27 commented:
Closing on the basis that while these features in ocamldebug are definitely useful, this is not an approach (stop-gap or otherwise) that's likely to make it over the line any time soon.

dra27 commented:
I'm pre-emptively locking conversation on this PR, to avoid any chance of a further AI pile-on. There are ongoing discussions on policy for AI-assisted contributions in #14383 and #14052.
